server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -685,10 +701,14 @@ static int getIndexNumberOfRoutingShards(Settings indexSettings, @Nullable Index
      * table based on the live nodes.
      */
     static ClusterState clusterStateCreateIndex(ClusterState currentState, Set<ClusterBlock> clusterBlocks, IndexMetadata indexMetadata,
-                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable) {
-        Metadata newMetadata = Metadata.builder(currentState.metadata())
-            .put(indexMetadata, false)
-            .build();
+                                                BiFunction<ClusterState, String, ClusterState> rerouteRoutingTable,
+                                                BiFunction<Metadata.Builder, IndexMetadata, Metadata.Builder> metadataTransaction) {
+        Metadata.Builder builder = Metadata.builder(currentState.metadata())
+            .put(indexMetadata, false);
+        if (metadataTransaction != null) {
+            builder = metadataTransaction.apply(builder, indexMetadata);
```

Comment:
Drive-by comment, but does this really need to return a brand new builder? Since the existing builder is being passed in couldn't this be a `BiConsumer<Metadata.Builder, IndexMetadata>` and modify the builder directly?

Reply:
That makes sense to me if the return value on builder methods is only for enabling a fluent API and not for changing the identity of the builder.

Reply:
This PR didn't have any uses of the actual transaction builder, what is the expected use that will change the identity of the builder?

Reply:
@dakrone, it's for rolling over data streams. Create the new backing index and add it to the data stream in the same cluster state update operation.

Reply:
I agree that there is no need need for the `metadataTransaction` to return a new builder instance.
Returning the same instance should suffice for the use case we would like to do with this change.

Reply:
maybe in that spirit also add an assertion that the same metadata builder instance is returned from the function?

Reply:
I think we should just turn it into a `BiConsumer` instead?